### ISSUE TITLE
Brush issues again

### ISSUE THREAD

**Issue by jhvanderven** [emeeks/semiotic#161]
I am revisiting the code from [this issue](https://github.com/emeeks/semiotic/issues/55) but now on a more recent version of semiotic (^1.7.13), d3-scale (^2.0.0) and react (^16.2.0). When I run the code as is, I get the following react error:
"Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops." The console adds:
"The above error occurred in the <XYFrame> component:
    in XYFrame (at TimeSelector.js:135)"

This is before any interaction has taken place. The page renders for a split second and then errors out.
If I comment out:
`          //xScaleType={scaleTime()}
`
the page renders, but the brush does not work as expected: the handles stay where they are and the ticks get updated very fast (as compared to the brush movement).

I have commented out the setState calls in brushDuring and brushEnd, but that did not change the behavior.

For completeness sake, here is my component:

```
import React from "react";
import { XYFrame } from "semiotic";
import { scaleTime } from "d3-scale";
var data;

export default class TimeSelector extends React.Component {

constructor(props) {
  super(props);
  data = [
    {
      px:this.props.start,py:1
    },
    {
      px:this.props.stop,py:1
    }
  ];
    
  this.state = {
    start:data[0].px,
    stop:data[1].px,
    data:data,
    align:true
  }

  this.brushDuring = this.brushDuring.bind(this);
  this.brushEnd = this.brushEnd.bind(this);
}

brushDuring(e) {
  if(e!==null && e!==undefined){
   this.setState({start:e[0], stop:e[1]});
  }
}

brushEnd(e) {
  if(e!==null && e!==undefined){
    this.setState({start:e[0], stop:e[1]});
  }
}

render() {
  return (
    <div>
      <div>
        <XYFrame
          className='-ts'
          // leave some room (60 pixels) for the controls
          size={[this.props.width, this.props.height-30]}
          lines={[{coordinates:this.state.data}]}
          xAccessor="px"
          yAccessor="py"
          //xScaleType={scaleTime()}
          yExtent={[0,2]}
          xExtent={[this.state.start, this.state.stop]}
          lineStyle={d => ({
            stroke: "grey",
            strokeWidth: "10px",
            opacity: 0.4,
          })}
          axes={[
            {
              orient: "bottom",
              ticks: 10
            }
          ]}
          margin={{ left: 60, top: 55, bottom: 40, right: 40 }}
          interaction={{
            during: this.brushDuring,
            end: this.brushEnd,
            brush: "xBrush",
            extent: [this.state.start,this.state.stop],
          }}
        />
      </div>
    </div>
    );
  }
};

```
I can see that the brush [example](https://emeeks.github.io/semiotic/#/semiotic/linebrush) still works, which puzzles me.

Update: If I remove the callbacks from the interaction, then the behaviour is as expected, but the events do not fire, of course.
```
          interaction={{
            //during: this.brushDuring,
            //end: this.brushEnd,
            brush: "xBrush",
            extent: [this.state.start,this.state.stop],
          }}

```
Update 2: If I comment out the following lines:
In the constructor:
```
//  this.brushDuring = this.brushDuring.bind(this);
//  this.brushEnd = this.brushEnd.bind(this);
```
In the render function:
`          //xScaleType={scaleTime()}
`
And redefine my event callbacks as arrow functions:
```
brushDuring = (e) => {
  if(e!==null && e!==undefined){
   this.setState({start:e[0], stop:e[1]});
  }
}

brushEnd = (e) => {
  if(e!==null && e!==undefined){
    this.setState({start:e[0], stop:e[1]});
  }
}

```

everything seems to work just fine.

**Comment by emeeks** [comment]
Looks like I'm not accounting for time scales in my function that checks to see if a brush has changed, I'll put together a fix for this.

**Comment by emeeks** [comment]
I notice you're setting the `xExtent` and the `extent` in your `interaction` for the brush to the same array. I'm not sure but this could be causing the problem as well. I added some code in Brush to explicitly deal with dates but looking at the existing example on the code it actually seems to be handling dates fine. I'll close this when I get 1.8 released and if it's still causing problems please feel free to re-open but I'll need to see the data driving this example to continue to look into it.

You can use https://codesandbox.io for a more involved example or a simple repo or gist, whichever is easiest for you.

**Comment by emeeks** [comment]
Okay 1.8.1 is out there. I'll leave this open let me know if you still see it.

**Comment by jhvanderven** [comment]
Updated semiotic. I still see the same behaviour. The data driving the example is in the code:
```
  data = [
    {
      px:this.props.start,py:1
    },
    {
      px:this.props.stop,py:1
    }
  ];
``` 

I am using the brush to select a time interval, the actual xyframe is a straight line. The frame drives the time interval of the other semiotic frames on the page. So, I pass the component a start and end time, which can then be altered, parent frame will receive the update (that's the part I left out) and then pass the changed interval to all kids.

I have removed the extent from the frame, not from the interaction. No improvement.

It looks like there's some sort of recursion going on when the xScaleType is set to datetime. 

Another part I left out, is the formatting of the ticks on the axes. What else would be the difference between a linear scale and a time scale?

Could this be the build environment again? Using create-react-app and npm...

**Comment by emeeks** [comment]
What I mean is the actual dataset itself. If you can give me the data that's being passed to `lines` as well as whatever is in the `props` or `state` of the component that is driving this example, that would help me to debug it. I have a feeling this is not a build problem but I can't reproduce it.

**Comment by jhvanderven** [comment]
The lines data is the dataset above, 2 points with the same y value.
Here is a screenshot of the component in action, the code above is a simpler version.

![screen shot 2018-03-02 at 12 10 08](https://user-images.githubusercontent.com/2427214/36896378-bd8f9e18-1e12-11e8-99b0-efffc6b7b3fc.png)

props.start is now - 14 days, and props.stop=now+14 days, width 1450, height 140.

The goal is to let the user select a time range, nothing else.

**Comment by emeeks** [comment]
I understand the principle, we use it in several places, the thing is I can't seem to recreate it. I added a two-point timeScale-based brush here:

https://emeeks.github.io/semiotic/#/semiotic/linebrush

And just don't see any issue with that approach. I think the bug is that you're continuously reseting the `xExtent` of the frame with the newly selected region. If your purpose is a time selector, you probably want the xExtent of the frame to stay fixed. Could you remove your `xExtent` setting and let me know if it continues to give you problems?

**Comment by jhvanderven** [comment]
I am a bit flabbergasted myself now. I tried to create a codesandbox.io to show you my error, but there I get the 'Cannot read property 'selection' of null' again, so that did not work. I then verified that I already removed the xExtent, put the timescale back in, and it worked. I think it was a restart of chrome that fixed it.

Now, if I put in the xExtent, it does not crash, but it shows the axes tickmarks in a wierd way instead of brushing. In fact, I cannot get it to crash at all.

I am sorry for the confusion. Before posting another issue, I will restart/reboot and see if the problem persists.
